spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,162 @@
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		Assert.isTrue(maxFetchSize > 0, "maxFetchSize has to be a positive integer.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+		super.setMaxFetchSize(maxFetchSize);
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		Assert.notNull(ackMode, "The acknowledgement mode can't be null.");
+		this.ackMode = ackMode;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		Assert.notNull(payloadType, "The payload type cannot be null.");
+		this.payloadType = payloadType;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		Assert.notNull(headerMapper, "The header mapper can't be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	/**
+	 * Instructs synchronous pull to wait until at least one message is available.
+	 */
+	public void blockOnPull() {
+		this.blockOnPull = true;
+	}
+
+	/**
+	 * Provides a single polled message.
+	 * <p>Messages are received from Pub/Sub by synchronous pull, in batches determined
+	 * by {@code fetchSize}.
+	 * <p>Cached messages in {@link AckMode#AUTO_ACK} or {@link AckMode#AUTO} are not
+	 * automatically acked until they are taken out of cache.
+	 * @param fetchSize number of messages to fetch from Pub/Sub.
+	 * @return {@link Message} wrapper containing the original message.
+	 */
+	@Override
+	protected Object doReceive(int fetchSize) {
+		if (this.cachedMessages.isEmpty()) {
+			List<ConvertedAcknowledgeablePubsubMessage> messages
+					= this.pubSubSubscriberOperations.pullAndConvert(this.subscriptionName, fetchSize, !this.blockOnPull, this.payloadType);
+			if (messages.isEmpty()) {
+				return null;
+			}
+			else if (messages.size() == 1) {
+				// don't bother storing.
+				return processMessage(messages.get(0));
+			}
+			else {
+				this.cachedMessages.addAll(messages);
+			}
+		}
+
+		return processMessage(this.cachedMessages.pollFirst());
+	}
+
+	@Override
+	public String getComponentType() {
+		return "gcp-pubsub:message-source";
+	}
+
+	/**
+	 * Applies header customizations and acknowledges the message, if necessary.
+	 * <p>{@link AckMode#AUTO} and {@link AckMode#AUTO_ACK} result in automatic acking on
+	 * success. {@link AckMode#AUTO} results in automatic nacking on failure.
+	 * @param message source Pub/Sub message.
+	 * @return {@link Message} wrapper containing the original message.
+	 */
+	private Message<Object> processMessage(ConvertedAcknowledgeablePubsubMessage message) {
+
+		Map<String, Object> messageHeaders =
+				this.headerMapper.toHeaders(message.getPubsubMessage().getAttributesMap());
+
+		if (this.ackMode == AckMode.MANUAL) {
+			// Send the original message downstream so user decides on when to ack/nack.
+			messageHeaders.put(GcpPubSubHeaders.ORIGINAL_MESSAGE, message);
```

Comment:
I think we need to revise this logic only in favor of `AcknowledgmentCallback` in the headers. 
See the same `AmqpMessageSource` or `KafkaMessageSource`